[Chai Jiexuan] iP

jiexuanc opened the pull request:
#### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)
DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html)
2. double-click it
3. add your tasks here
4. let it manage your tasks for you :+1:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jiexuanc committed:
Implemented skeletal version of Duke.
Echo commands and exit on "bye" command

jiexuanc committed:
Implemented Add and List.
Created DukeList class to encapsulate the list structure

jiexuanc committed:
Removed DukeList.java, shifted all functionality to Duke
Added Task to encapsulate individual task
Added Mark/Unmark functionality

jiexuanc committed:
Added ToDos, Deadlines and Events

jiexuanc committed:
Changed to ArrayList

jiexuanc committed:
Added borders to Duke
Added input and expected results for automated

jiexuanc committed:
Fixed bugs.
Added Error handling for most inputs.

jiexuanc committed:
Changed test input/output

jiexuanc committed:
Added Delete function
Update test cases for automated testing.

jiexuanc committed:
Modified error message and implementation.

jiexuanc committed:
Updated Duke implementation

jiexuanc committed:
Fix bugs

jiexuanc committed:
Deadline class: fix error handling bugs

jiexuanc committed:
Duke class: Fix string formatting bug

jiexuanc committed:
Add javadocs and comments
Change most attributes/methods to suitable access modifier
Fix Duke bug where program does not terminate

jiexuanc committed:
Add Storage class to support saving tasks to hard disk

Tasks are refreshed when Duke starts up. No information can be stored.

Retrieving past tasks should be supported.

Add Storage class to create and manage files storing information of the tasks. Initializes Duke with the initial state of the tasks as per the file.

jiexuanc committed:
Add Storage class to support saving tasks to hard disk

Tasks are refreshed when Duke starts up. No information can be stored.

Retrieving past tasks should be supported.

Add Storage class to create and manage files storing information of the tasks. Initializes Duke with the initial state of the tasks as per the file.

jiexuanc committed:
Add Parser class to handle date and time

Duke initially treats all time frame as String only.

Supports date and time for greater utility.

jiexuanc committed:
Add Parser class to handle date and time

Duke initially treats all time frame as String only.

Supports date and time for greater utility.

jiexuanc committed:
Add TaskList class to abstract task manager

jiexuanc committed:
Add Command class to abstract each command execution

jiexuanc committed:
Add Parser class to parse user inputs

All user inputs were handled by Duke class. Code is messy and difficult to debug

Creates Parser class to manage user inputs and handle errors, also supports formatting of date/time

jiexuanc committed:
Fix bugs regarding exception handling of user inputs

jiexuanc committed:
Add MaybeDate class to abstract the time frame of tasks

Events with time frame either only support Strings or only support LocalDateTime

Support both versions by creating MaybeDate and modify other dependent classes to support the change.

jiexuanc committed:
Update Ui class to deal with interactions with users

Ui class is not functional and each class interacted with the user by themselves

Ui class allows interactions with users to be centralized at one point. Easier management of code.

Let's move all print functions from other classes to Ui.

jiexuanc committed:
Merge branch 'add-gradle-support'

jiexuanc committed:
Add JUnit test with Gradle support

Manual unit testing is inefficient.

Automate testing with JUnit.

jiexuanc committed:
Move classes into duke package

jiexuanc committed:
Fix style errors

Ignored some warning from checkstyle.

jiexuanc committed:
Add JavaDocs.

jiexuanc committed:
Follows coding standard

jiexuanc committed:
Add find function to TaskList class

Allows Duke to search for tasks based on keyword.

jiexuanc committed:
Merge branch 'branch-A-JavaDoc'

jiexuanc committed:
Follows coding standard

jiexuanc committed:
Merge branch 'branch-A-CodingStandard'

jiexuanc committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Command.java
#	src/main/java/duke/TaskList.java

jiexuanc committed:
Adds JavaDoc for Parser class

jiexuanc committed:
Fix bugs

jiexuanc committed:
Adds help message

jiexuanc committed:
Adds help message

jiexuanc committed:
Updates README.md

jiexuanc committed:
Updates build.gradle to support jar

jiexuanc committed:
Change chat-bot name

jiexuanc committed:
Change chat-bot name

yaojiax commented on src/main/java/elise/MaybeDate.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Abstraction of a period represented by either String or date and time.
+ */
+public class MaybeDate {
```
I like how you abstracted a class for the Date which can be either of type String or a LocalDate. Very creative 👍

yaojiax commented on src/main/java/elise/Parser.java:
```diff
@@ -0,0 +1,110 @@
+        if (s.length() > 500) {
+            throw new EliseException("Message body is too long!");
+        }
+        switch (command) {
```
Good that you practised the correct indentation for switch case, as it is a common mistake. Keep it up

yaojiax commented on src/main/java/elise/Ui.java:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * Prints help message.
+     */
+    protected void showHelp() {
```
You could possibly use a PrintWriter class here to print these many statements as the current implementation with System.out.println incurs significant OS overhead, so a PrintWriter can achieve better performance results. But otherwise, great feature 👍

yaojiax commented on src/main/java/elise/Command.java:
```diff
@@ -0,0 +1,134 @@
+ * Stores all information necessary for execution.
+ */
+public class Command {
+    /*
```
This comment helped me understand the code faster! Nice!

yaojiax commented on src/main/java/elise/ToDo.java:
```diff
@@ -0,0 +1,36 @@
+     * @param status Completed or not.
+     * @param content
+     */
+    protected ToDo(boolean status, String[] content) {
```
Good use of protected here and elsewhere in your code. 👍

yaojiax reviewed (commented):
Overall, wonderful code that follows OOP principles with basically no coding style violation!

jiexuanc committed:
Parser class: replace with enum constants

chunzkok commented on src/main/java/elise/Elise.java:
```diff
@@ -0,0 +1,59 @@
+                }
+            } catch (EliseException e) {
+                ui.showError(e);
+            }
```
I like how clean this code is! All thanks to the effort put in in factoring out all the commands 👍

chunzkok commented on src/main/java/elise/Command.java:
```diff
@@ -0,0 +1,134 @@
+    6 - event
+    7 - delete
+    8 - find
+     */
```
Looks like the perfect use case for `Enum`s! perhaps you could reuse those present in Parser.java?

chunzkok commented on src/main/java/elise/Command.java:
```diff
@@ -0,0 +1,134 @@
+     * @throws EliseException EliseException unique to Elise
+     * @throws IOException Unexpected IOException
+     */
+    protected void execute(Ui ui, TaskList taskList, Storage storage) throws EliseException, IOException {
```
Would probably be quite a bit neater if these are split into separate subclasses of `Command` that override this `execute` method to make use of polymorphism!
However, that will definitely take quite a bit of precious time 😭 so I understand why it might not be of high priority

chunzkok commented on src/main/java/elise/Parser.java:
```diff
@@ -0,0 +1,122 @@
+            command = Word.valueOf(sc.next().toUpperCase());
+        } catch (IllegalArgumentException e) {
+            throw new EliseException("Invalid input.");
+        }
```
Good idea! now the commands are case-insensitive YAY

chunzkok commented on src/main/java/elise/Task.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Represents a task with modifiable status.
+ */
+public abstract class Task {
```
nice! I like how this is abstract to ensure we only instantiate concrete implementations like `ToDo`, `Deadline` etc.

chunzkok commented on src/main/java/elise/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+     * Lists out all tasks and their status.
+     */
+    protected void list() {
+        if (tasks.size() == 0) {
```
Nice, good to let users know the list is empty instead of displaying nothing when looping through an empty list 😩

chunzkok commented on src/main/java/elise/Ui.java:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * Prints help message.
+     */
+    protected void showHelp() {
```
Wow!! This is amazing, it is probably one of the most useful features a user can use especially if they're new 👍 💯

chunzkok reviewed (commented):
Thanks for (involuntarily??) sharing your code! I've learnt a lot from your input validation cases as well as the special help command 😄 Can't wait to see the final version of Elise!!!

jiexuanc committed:
GUI Changes

jiexuanc committed:
Implement initial version of GUI

jiexuanc committed:
Fix cutting-off bug of GUI

jiexuanc committed:
Add exit function on command

jiexuanc committed:
Add GUI with JavaFX

Only command line interface is present.

A graphical user interface is more user-friendly because it is
more visually appealing.

Let's implement a simple GUI with JavaFX.

Future change to design of GUI is to be expected.

jiexuanc committed:
Fix bugs for help command

jiexuanc committed:
Add Launcher class to workaround classpath

jiexuanc committed:
Fix GUI alignment

jiexuanc committed:
Fix GUI alignment

jiexuanc committed:
Add README.md

jiexuanc committed:
Refactor code

jiexuanc committed:
Create DateParser class for more flexible date format

Date format is restrictive.

A more flexible date format is more user-friendly because users can enter shorter messages and do not need to remember the original format

Let's
* create more date formats
* allow for case-insensitive input

jiexuanc committed:
Separate commands, tasks and internals into packages

All classes were in elise package.

Separating into individual packages improves the code quality.

As a step towards improving code quality, let's move related classes to their respective packages.

jiexuanc committed:
Merge pull request #2 from jiexuanc/branch-A-CodeQuality

Separate commands, tasks and internals into packages

jiexuanc committed:
Separate commands, tasks and internals into packages

All classes were in elise package.

Separating into individual packages improves the code quality.

As a step towards improving code quality, let's move related classes to their respective packages.

jiexuanc committed:
Merge pull request #3 from jiexuanc/branch-C-NaturalDates

Create DateParser class for more flexible date format

jiexuanc committed:
Refactor code

jiexuanc committed:
Refactor code

jiexuanc committed:
Fix style violations

jiexuanc committed:
Add JavaDocs and JUnit tests

jiexuanc committed:
Add Java asserts

jiexuanc committed:
Add Java asserts

jiexuanc committed:
Merge pull request #4 from jiexuanc/branch-A-Assertions

Add Java asserts

jiexuanc committed:
Modify GUI

jiexuanc committed:
Edit help message

jiexuanc committed:
Add README.md

jiexuanc committed:
Add config file for git page

jiexuanc committed:
Add Ui.png for product showcase

jiexuanc committed:
Update _config.yml

jiexuanc committed:
Update _config.yml

jiexuanc committed:
Edit README.md

jiexuanc committed:
Edit README.md

jiexuanc committed:
Update README.md

jiexuanc committed:
Update README.md
